src/pool.rs

```diff
@@ -14,9 +14,25 @@ const DEFAULT_MAX_IDLE_CONNECTIONS_PER_HOST: usize = 1;
 
 /// Holder of recycled connections.
 ///
-/// Invariant: The length of recycle and lru are the same.
-/// Invariant: Each PoolKey exists as a key in recycle, and vice versa.
-/// Invariant: Each PoolKey exists in recycle at most once and lru at most once.
+/// For each PoolKey (approximately hostname and port), there may be
+/// multiple connections stored in the `recycle` map. If so, they are stored in
+/// order from oldest at the front to freshest at the back.
+///
+/// The `lru` VecDeque is a companion struct to `recycle`, and is used to keep
+/// track of which connections to expire if the pool is full on the next insert.
+/// A given PoolKey can occur in lru multiple times. The first entry in lru for
+/// a key K represents the first entry in `recycle[K]`. The second entry in lru
+/// for `K` represents the second entry in `recycle[K]`, and so on. In other
+/// words, `lru` is ordered the same way as the VecDeque entries in `recycle`:
+/// oldest at the front, freshest at the back. This allows keeping track of which
+/// host should have its connection dropped next.
+///
+/// These invariants hold at the start and end of each method:
+///  - The length `lru` is equal to the sum of lengths of `recycle`'s VecDeques.
+///  - Each PoolKey exists the same number of times in `lru` as it has entries in `recycle`.
+///  - If there is an entry in `recycle`, the VecDeque has at least one element.
```

Comment:
> the VecDeque has at least one element

Since both are `VecDeque`, this is a bit ambiguous. Maybe call it `lru` here?

Reply:
Good call. It turns out this was confusing, since I actually meant to refer to the VecDeque that is a member of `recycle`, not `lru`. Changed to "If there is an entry in `recycle`, it has at least one element".